### PR TITLE
Gnome and Pantheon: install nixos wallpapers

### LINKED ISSUE #86146: Default NixOS wallpaper missing after update
**Describe the bug**
The default wallpaper was a NixOS flake. I actually like it and never changed it. Now i miss it.

Screenshot from 2020-04-27 23-22-13

**To Reproduce**
1. use pantheon desktop
2. update to 20.03
3. login

**Expected behavior**
same default wallpaper as before instead of black screen

**Notify maintainers**
@worldofpeace 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.35, NixOS, 20.03.1589.9137f05564e (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - channels(davidak): `""`
 - channels(root): `"nixos-20.03.1589.9137f05564e, nixos-hardware, nixos-unstable-20.09pre222973.7c399a4ee08"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Fix https://github.com/NixOS/nixpkgs/issues/86146.
We were hardcoding them into settings but not installing them.

This also changes nixos-artwork slightly https://github.com/NixOS/nixpkgs/commit/581122f1a99af0f4ced14db686726d64f358ccb6. Namely, install to share/backgrounds, and provide an absolute path to the file.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/misc/nixos-artwork/wallpapers.nix [hunk 1]
```diff
@@ -3,12 +3,18 @@
...
 
       dontUnpack = true;
 
       installPhase = ''
+        # GNOME
+        mkdir -p $out/share/backgrounds/nixos
+        ln -s $src $out/share/backgrounds/nixos/${src.name}
+
+        # TODO: is this path still needed?
         mkdir -p $out/share/artwork/gnome
```
Comment (jtojnar) [inline pkgs/data/misc/nixos-artwork/wallpapers.nix:18]: Do not see `artwork` used within path when searching g-c-c. Googling "/usr/share/artwork/gnome" shows no results, without `gnome` only few mentions but none of them look official. We might want to keep this for backwards compatibility, though.

**Comment by worldofpeace** [comment]
I don't really understand the failing check here.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg eval

(I have no idea)

**Comment by jtojnar** [comment]
Would using `builtins.toString` on the path remove the context and fix the error?

**Comment by worldofpeace** [comment]
> Would using `builtins.toString` on the path remove the context and fix the error?

Nope, I tried that at first and still got
```
error: in 'toFile': the file 'options.xml' cannot refer to derivation outputs, at /home/worldofpeace/Code/nix/nixpkgs/nixos/lib/make-options-doc/default.nix:89:16
```

when making the manual.

**Comment by jtojnar** [comment]
Using the following patch,

```patch
--- a/nixos/modules/services/x11/display-managers/lightdm.nix
+++ b/nixos/modules/services/x11/display-managers/lightdm.nix
@@ -133,7 +133,7 @@ in
 
       background = mkOption {
         type = types.path;
-        default = pkgs.nixos-artwork.wallpapers.simple-dark-gray-bottom.gnomeFilePath;
+        default = let pkg = pkgs.nixos-artwork.wallpapers.simple-dark-gray-bottom; pp = "${pkg}/share/backgrounds/nixos/${pkg.src.name}"; in builtins.trace (builtins.getContext pkg.gnomeFilePath) (builtins.trace (builtins.getContext pp) pp);
         description = ''
           The background image or color to use.
         '';
```

it looks like the issue is actually the opposite than I thought:

```
trace: { "/nix/store/qry13wjmx6wbj4rcb6qz2yis4m3bhin5-simple-dark-gray-2018-08-28.drv" = { outputs = [ "out" ]; }; }
trace: { }
```

Even though the both strings are the same, `passthru` strips the context, so `toFile` sees a path that was not mentioned in any context and complains.

**Comment by jtojnar** [comment]
This patch seems to work:

```patch
--- a/nixos/modules/services/x11/display-managers/lightdm.nix
+++ b/nixos/modules/services/x11/display-managers/lightdm.nix
@@ -133,9 +133,7 @@ in
 
       background = mkOption {
         type = types.path;
-        # TODO: use gnomeFilePath when fixing string context issues in make-option-doc
-        # https://github.com/NixOS/nixpkgs/issues/83863
-        default = "${pkgs.nixos-artwork.wallpapers.simple-dark-gray-bottom}/share/artwork/gnome/nix-wallpaper-simple-dark-gray_bottom.png";
+        default = pkgs.nixos-artwork.wallpapers.simple-dark-gray-bottom.gnomeFilePath;
         description = ''
           The background image or color to use.
         '';
--- a/pkgs/data/misc/nixos-artwork/wallpapers.nix
+++ b/pkgs/data/misc/nixos-artwork/wallpapers.nix
@@ -29,7 +29,7 @@ _EOF
       '';
 
       passthru = {
-        gnomeFilePath = "${pkg}/share/backgrounds/nixos/${src.name}";
+        gnomeFilePath = builtins.unsafeDiscardStringContext "${pkg}/share/backgrounds/nixos/${src.name}";
         kdeFilePath = "/share/wallpapers/${name}/contents/images/${src.name}";
       };
 
```

Obviously, not a good idea. Not only it is unsafe but it would also break dependence tracking when used in derivations.

I am not sure why does not [`builtins.toFile`](https://nixos.org/nix/manual/#builtin-toFile) support referencing the result of derivation, ¿perhaps it happens in evaluation time so it would cause IFD?

What is even weirder, when we construct the string in place as before, it shows no context:

```nix
trace: { }
```

but if we do the same in repl, we get a context:

```nix
nix-repl> let pkg = pkgs.nixos-artwork.wallpapers.simple-dark-gray-bottom; pp = "${pkg}/share/backgrounds/nixos/${pkg.src.name}"; in builtins.trace (builtins.getContext (builtins.toString pp)) pp
trace: { "/nix/store/qry13wjmx6wbj4rcb6qz2yis4m3bhin5-simple-dark-gray-2018-08-28.drv" = { outputs = [ "out" ]; }; }
"/nix/store/pzgmk5wx85drxbw1ikyg50gc1whi36ag-simple-dark-gray-2018-08-28/share/backgrounds/nixos/nix-wallpaper-simple-dark-gray_bottom.png"
```

**Comment by jtojnar** [comment]
It makes sense for the manual not to depend on other attributes but at the same time user’s configuration should. Not sure if this is solvable.

**Comment by jtojnar** [comment]
Hmm, `mkOption` actually supports `defaultText` so we can have something like a cake and eat it too. The downside is that it is rendered like a string, not a literal:

image

**Comment by worldofpeace** [comment]
@jtojnar Yay :sparkles:. That seems fine to me. Re the investigation https://github.com/NixOS/nixpkgs/pull/86163#issuecomment-620981858, perhaps @edolstra can shed some light.

**Pull request merged**